File: 045.py
def tri(n):
    return n*(n+1)/2

def penta(n):
    return n*(3*n - 1)/2
    
def hexa(n):
    return n*(2*n - 1)

    
t = 285
p = 165
h = 143

t += 1

tt = tri(t)
pp = penta(p)
hh = hexa(h)


while True:
    if tt == pp == hh:
        break
    if tt < pp and tt < hh:
        t += 1
        tt = tri(t)
    if pp < tt and pp < hh:
        p += 1
        pp = penta(p)
    if hh < tt and hh < pp:
        h += 1
        hh = hexa(hh)
    
print tt

File: 045.cpp
#include <stdio.h>

int tri(int n){
    return n*(n+1)/2;
}

int penta(int n){
    return n*(3*n - 1)/2;
}
    
int hexa(int n){
    return n*(2*n - 1);
}


int main(){
    
    int t = 285;
    int p = 165;
    int h = 143;

    t++;

    int tt = tri(t);
    int pp = penta(p);
    int hh = hexa(h);

    
    while (1) {
        if (tt == pp && tt == hh) {
            break;
        }
        
        while (tt < pp && tt < hh) {
            t++;
            tt = tri(t);
        }
        while (pp < tt && pp < hh) {
            p++;
            pp = penta(p);
        }
        while (hh < tt && hh < pp) {
            h++;
            hh = hexa(h);
        }
    }
    
    printf("tt: %d\n", tt);
    
    return 0;
}
